File: hacker-rank-interview-preparation-kit/Ruby/dictionaries-and-hashmaps/two_strings.rb
=begin
Tae H0 K1m
  11. Two Strings

  Given two strings, determine if they share a common substring. A substring
  may be as small as one character.

  For example, the words "a", "and", "art" share the common substring . The
  words "be" and "cat" do not share a substring.

  it should return a string, either YES or NO based on whether the strings
  share a common substring.
=end

# 1. store characters frequency in first word
# 2. check if the second word has the common characters
# 3. if there is, print "YES", otherwise print "NO"

def twoStrings(s1, s2)
  freq = Hash.new

  # 1) store characters frequency in first word
  for i in 0...(s1.length)
    freq[s1[i]] = 1
  end

  # 2) check if the second word has the common characters
  for i in 0...(s2.length)
    return "YES" unless freq[s2[i]].nil?
  end

  return "NO"
end

puts twoStrings("hello", "world")
# => YES

puts twoStrings("hi", "world")
# => NO

File: hacker-rank-interview-preparation-kit/Python/dictionaries-and-hashmaps/two_strings.py
"""
    11. Two Strings
    Tae H0 K1m

    problem: Determine if two strings share a common substring. A substring
    maybe as small as one character. Return either "Yes" or "No" based on
    whehter the strings share a common substing

    input: two strings
    output: "Yes" or "No " based on whehter the strings share a common substing

    approach:
        0. check if the two strings share a character
        1. store frequency of character in the first word
        2. iterate through the second word
        3. if the current character is in the first word, return "Yes"
        4. after iteration, return "No"

    example:

    hello, world => Yes
    hi, world => No

"""

def twoStrings(s1, s2):
    freq = {}
    for char in s1:
        if char not in freq:
            freq[char] = 1

    for char in s2:
        if char in freq:
            return "Yes"

    return "No"

print(twoStrings("hello", "world"))
# => Yes

print(twoStrings("hi", "world"))
# => No
